htmd/qm/custom.py

```diff
@@ -118,18 +114,27 @@ def minimize(self, coords, restrained_dihedrals=None):
             fidx = self.system.addForce(f)
             forceidx.append(fidx)
 
-        integrator = mm.LangevinIntegrator(0, 0, 0)
-        self.sim = app.Simulation(self.structure.topology, self.system, integrator, self.platform, self.platprop)
-
         if coords.ndim == 3:
             coords = coords[:, :, 0]
-        self.sim.context.setPositions(coords * unit.angstrom)
-        self.sim.minimizeEnergy()
-        state = self.sim.context.getState(getEnergy=True, getPositions=True)
-        endcoords = state.getPositions(asNumpy=True).value_in_unit(unit.angstrom)
+
+        natoms = coords.shape[0]
+
+        integrator = mm.LangevinIntegrator(0, 0, 0)
+        sim = app.Simulation(self.structure.topology, self.system, integrator, self.platform, self.platprop)
+
+        def goalFunc(x):
+            sim.context.setPositions(x.reshape((natoms, 3)) * unit.angstrom)
+            state = sim.context.getState(getEnergy=True, getForces=True)
+            energy = state.getPotentialEnergy().value_in_unit(unit.kilocalories_per_mole)
+            forces = state.getForces(asNumpy=True).value_in_unit(unit.kilocalories_per_mole / unit.angstrom)
+            grad = -forces.reshape(-1)
+            return energy, grad
+
+        res = minimize(goalFunc, coords.reshape(-1), method='L-BFGS-B', jac=True, options={'ftol': 0, 'gtol': 0.05})
```

Comment:
How have you the selected tolerances. It would be good to document with a comment.

Reply:
https://wiki.fysik.dtu.dk/ase/ase/optimize.html
I copied the tolerance from the ASE tutorial which is as arbitrary as the other thresholds we choose I would say. We don't document the 1e-5 tolerance for fitting either. As to why force tolerance, scipy doesn't support absolute energy tolerances, and it makes sense to use force tolerance for minimization.

Reply:
I tried also the relative 1 kjoule/mol used in OpenMM but it seemed to finish too early on my tests without reaching a good minimum.

Reply:
In what units is `gtol`? kcal/mol/A or kJ/mol/A?

Reply:
OK! It is tighter than Psi4 max force tolerance.